### PR TITLE
Document opt_ API

### PR DESCRIPTION
**Pull request by richsalz**
Also made some functions static that didn't need to be public.

### CODE REVIEW

File: doc/internal/man3/opt.pod [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ void opt_help(const OPTIONS *list);
+ int opt_isdir(const char *name);
+
+=head1 DESCRIPTION
+
+The functions on this page provide a common set of option-parsing for
+the OpenSSL command and the internal test programs.
+It is intended to be used like the standard getopt(3) routine, except
+that multi-character flag names are supported, and a variety of parsing
+and other utilities functions are also provided.
```
Comment (slontis) [inline doc/internal/man3/opt.pod:59]: utility ?

File: doc/internal/man3/opt.pod [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    typedef enum OPTION_choice {
+        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+        OPT_YES, OPT_NAME, OPT_COUNT, OPT_OFILE
+    } OPTION_CHOICE;
+
+The first two lines must appear exactly as shown. The third line is a sample
+set of flags, and the closing C<typedef> name is used for error-checking
+as discussed below.
+
+The B<OPTIONS> C<typedef> specifies an options, what type of argument
```
Comment (slontis) [inline doc/internal/man3/opt.pod:75]: an = any

File: doc/internal/man3/opt.pod [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    int retval;
+    int valtype;
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
+are words preceeded with a minus sign. If the user uses two minus signs,
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is a the value to return if the option is found. It should be
```
Comment (slontis) [inline doc/internal/man3/opt.pod:89]: remove 'a'

File: doc/internal/man3/opt.pod [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    's' A text string
+    '/' A directory
+    '<' Name of file to open for input
+    '>' Name of file to open for output
+    'n' A signed number that fits in the C<int> type
+    'p' A non-negative number that fits in the C<int> type
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
+    'l' A signed number that fits in the C<long> type
+    'u' Aunsigned number that fits in the C<unsigned long> type
```
Comment (slontis) [inline doc/internal/man3/opt.pod:106]: A unsigned

File: doc/internal/man3/opt.pod [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    \0  No value
+    '-' No value
+    's' A text string
+    '/' A directory
+    '<' Name of file to open for input
+    '>' Name of file to open for output
+    'n' A signed number that fits in the C<int> type
+    'p' A non-negative number that fits in the C<int> type
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
```
Comment (slontis) [inline doc/internal/man3/opt.pod:104]: Choose 'An' or 'A' for these

File: doc/internal/man3/opt.pod [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        {"output" OPT_OFILE, '>', "Output file; default is stdout"},
+        {NULL},
+    };
+
+Note that the B<OPT_HELP> option is explicitly listed, and the list ends with
+an entry of all-null's. The other two special options, B<OPT_ERR> and B<OPT_EOF>
+should not appear in the array.
+
+If the help string is too long to fit into one line, it may be continued
+on multiple lines each entry should use B<OPT_MORE_STR>, like this:
```
Comment (slontis) [inline doc/internal/man3/opt.pod:130]: , each

File: doc/internal/man3/opt.pod [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+        {"output" OPT_OFILE, '>', "Output file; default is stdout"},
+        {OPT_MORE_STR, 0, 0,
+         "This flag is not really needed on Unix systems"}
+        {OPT_MORE_STR, 0, 0,
+
+Each subsequent line will be indent the correct amount.
```
Comment (slontis) [inline doc/internal/man3/opt.pod:137]: indented?

File: doc/internal/man3/opt.pod [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+OPT_PARAM_STR, opt_progname, opt_appname, opt_getprog, opt_init, opt_format,
+format2str, opt_int, opt_long, opt_imax, opt_umax, opt_ulong, opt_pair,
+opt_string, opt_cipher, opt_md, opt_next, opt_arg, opt_flag, opt_unknown,
+opt_num_rest, opt_rest, opt_help, opt_isdir, opt_progname opt_next
+
+- Option parsing for commands and tests
+
+=head1 SYNOPSIS
+
+ #include "apps/include/opt.h"
```
Comment (levitte) [inline doc/internal/man3/opt.pod:14]: We never include it like that.  Rather, we compile with `-I PATH/TO/apps/include` and then have `#include "opt.h"`

File: doc/internal/man3/opt.pod (hunk 5 above)
Comment (mattcaswell) [reply doc/internal/man3/opt.pod:104]: Errrr....no? What @richsalz has here is correct. "An unsigned" but "A" for everything else.

**Comment by richsalz** [comment]
Thanks @slontis and @levitte .  Fixup commit pushed addressing your feedback. @mattcaswell was right about me being right on the global A/An change :)  I just marked things as resolved to reduce the email.  Now to finish it.

**Comment by richsalz** [comment]
Ready for review now. @ddvo, sorry about the "not apps/cms.c" comment, but it *is* true.

File: doc/internal/man3/OPTIONS.pod [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+Each program should define, near the main() routine, an enumeration
+that is the set of options the program accepts. For example:
+
+    typedef enum OPTION_choice {
+        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+        OPT_YES, OPT_NAME, OPT_COUNT, OPT_OFILE
+    } OPTION_CHOICE;
+
+The first two lines must appear exactly as shown. The third line is a sample
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:72]: Good that this is documented here.
Yet recently I have been wondering why each app is expected to define 
```
        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
which adds just redundancy and the risk of inconsistency.

File: doc/internal/man3/OPTIONS.pod [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    int valtype;
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
+are words preceeded with a minus sign. If the user uses two minus signs,
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is the value to return if the option is found. It should be
+one of the choices in the enumeration above.
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:91]: For the retvals it should also be mentioned that they are expected to be unique (though I'm going to add an exception to this rule soon) and the uniqueness is checked at runtime.

File: doc/internal/man3/OPTIONS.pod [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+be chosen from the following set:
+
+    \0  No value
+    '-' No value
+    's' A text string
+    '/' A directory
+    '<' Name of file to open for input
+    '>' Name of file to open for output
+    'n' A signed number that fits in the C<int> type
+    'p' A nonnegative number that fits in the C<int> type
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:103]: nonnegative -> positive

File: doc/internal/man3/OPTIONS.pod [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
+    'l' A signed number that fits in the C<long> type
+    'u' An unsigned number that fits in the C<unsigned long> type
+    'c' File of fomer PEM, DER, or S/MIME
+    'F' A file in PEM or DER format
+    'E' Like 'F' but also allows ENGINE
+    'f' Any file format
+
+The B<helpstr> is what to display when the user uses the help flag,
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:113]: "help flag" or "help option"?

File: doc/internal/man3/OPTIONS.pod (hunk 9 above)
Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod:72]: The `opt_next` function returns -1 on error and 0 when done.  Therefore both those values should be in the enumeration of possible values that the application is expecting.  You can cast the `opt_next` value to your enumeration type, and the compiler can warn you if you're missing cases.  That's done in the `OPT_x_CASES` code handling.

File: doc/internal/man3/OPTIONS.pod [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+which should be C<"help">.
+
+A program should declare its options right after the enumeration:
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:116]: Please mention that the order of the entries should be the same as in the enumeration, which helps readability and maintainability.

File: doc/internal/man3/OPTIONS.pod (hunk 10 above)
Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod:91]: No, the retval does not have to be unique.  It's how we alias options.  The *name* of the option has to be unique.
For example, the "enc" command  has
```
    {"a", OPT_A, '-', "Base64 encode/decode, depending on encryption flag"},
    {"base64", OPT_A, '-', "Same as option -a"},
```

I assume you're doing something like that.

File: doc/internal/man3/OPTIONS.pod [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+will also do range-check on the argument if the appropriate B<valtype>
+field is specified for the option.
+
+The opt_isdir() function returns B<1> if the specified B<name> is
+a directory, or B<0> if not. The opt_format() takes a string value,
+such as used in with the B<-informat> or related flags, and fills
+the value from the constants in F<fmt.h> file.
+
+The opt_string() function checks that B<name> appears in the
+B<NULL>-terminated array of strings. It returns B<1>ne if found,
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:236]: Funny style: `B<1>ne`

File: doc/internal/man3/OPTIONS.pod [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+After processing all the options, the opt_num_rest() returns what is
+left, and opt_rest() returns a pointer to the first non-option.
+If there were no parameters, it will point to the B<NULL> that is
+at the end of the standard B<argv> array.
+
+=head2 Example
+
+The best examples can be found in sources for the commands in the F<apps>
+directory of the source tree.
+A notable exception is F<apps/cmp.c> which uses this API, but does
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:253]: No problem mentioning this.
This CMP app allows putting all options also in a config file in addition to the command line, while the options on the command line can override those given in the config file.

**Review by DDvO — COMMENTED** [review]
Nice work.
I just added some quick comments.

File: doc/internal/man3/OPTIONS.pod (hunk 10 above)
Comment (DDvO) [reply doc/internal/man3/OPTIONS.pod:97]: Ah, I confused the two, so my comment was actually about the option names.
These are checked for uniqueness.

File: doc/internal/man3/OPTIONS.pod [hunk 16]
```diff
@@ -246,16 +254,41 @@ left, and opt_rest() returns a pointer to the first non-option.
...
+letter:
+
+    V   Validation
+    X   Extended certificate
+    S   TLS/SSL
+    R   Random state
+
+The B<OPT_x_ENUM> macro is used to define the numeration values, where B<x>
+is one of the letters above.  The B<OPT_x_OPTIONS> macro is used to
+list the set of common options, and the B<OPT_R_CASES> is used in
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:271]: OPT_x_CASES

Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod]: fixup commit pushed, thanks.

File: doc/internal/man3/OPTIONS.pod [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+the equivalent of B<stderr>. This is already done in the OpenSSL
+application.
+
+=head2 Data Types
+
+Each program should define, near the main() routine, an enumeration
+that is the set of options the program accepts. For example:
+
+    typedef enum OPTION_choice {
+        OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
Comment (beldmit) [inline doc/internal/man3/OPTIONS.pod:68]: If this set of options must appear exactly this, doesn't it mean that it's worth converting them to smth like verification options?

Comment (DDvO) [reply doc/internal/man3/OPTIONS.pod:68]: A very simple solution would be to turn
```
         OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
```
into a macro.

Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod:68]: Turning those common options into a macro is for another PR,.  This is purely documentation.  FWIW, I disagree with that idea and won't be doing it :)

Comment (DDvO) [reply doc/internal/man3/OPTIONS.pod:68]: That's why I wrote: would ;-)

**Comment by richsalz** [comment]
fixup: non-negative -> nonnegative

File: doc/internal/man3/OPTIONS.pod [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+it takes (if any), and an optional "help" string.  It is a C<struct>
+containing these fields:
+
+    const char *name;
+    int retval;
+    int valtype;
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
+are words preceeded with a minus sign. If the user uses two minus signs,
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:92]: preceeded -> preceded

File: doc/internal/man3/OPTIONS.pod [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+    '<' Name of file to open for input
+    '>' Name of file to open for output
+    'n' A signed number that fits in the C<int> type
+    'p' A positive number that fits in the C<int> type
+    'N' A nonnegative number that fits in the C<int> type
+    'M' A signed number that fits in the C<intmax_t> type
+    'U' An unsigned number that fits in the C<uintmax_t> type
+    'l' A signed number that fits in the C<long> type
+    'u' An unsigned number that fits in the C<unsigned long> type
+    'c' File of fomer PEM, DER, or S/MIME
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:115]: former

File: doc/internal/man3/OPTIONS.pod [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+This is a "wildcard" or "unknown" option. For example, it allows an
+option to be be taken as digest algorithm, like B<-sha1>. The
+function opt_cipher() takes the specified B<name> and fills in
+the cipher into B<cipherp>.  The function opt_md() does the same
+thing for message digest.
+
+There are a several useful functions for parsing numbers.  These are
+opt_int(), opt_long(), opt_ulong(), opt_imax(), and opt_umax().  They all
+take C<0x> to mean hexadecimal and C<0> to mean octal, and will do the
+necessary range-checking. They return B<1> if succcessful and fill in the
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:234]: successful

File: doc/internal/man3/OPTIONS.pod [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+
+OPTIONS, OPT_PAIR,
+opt_progname, opt_appname, opt_getprog, opt_init, opt_format,
+opt_int, opt_long, opt_imax, opt_umax, opt_ulong, opt_pair,
+opt_string, opt_cipher, opt_md, opt_next, opt_arg, opt_flag, opt_unknown,
+opt_num_rest, opt_rest, opt_help, opt_isdir
+- Option parsing for commands and tests
+
+=head1 SYNOPSIS
+
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:13]: The following symbols are not (yet) documented:
```
const char *opt_path_end(const char *filename);
void opt_begin(void);
const char *format2str(int format);
int opt_int_arg(void);
char *opt_name(void);
int opt_verify(int i, X509_VERIFY_PARAM *vpm);
int opt_rand(int i);
int opt_provider(int i);
void opt_print(const OPTIONS * opt, int doingparams, int width);
int opt_format_error(const char *s, unsigned long flags);
void print_format_error(int format, unsigned long flags);
int opt_printf_stderr(const char *fmt, ...);
```

**Review by DDvO — COMMENTED** [review]
A couple of typos and a bunch of missing functions

**Comment by DDvO** [comment]
Ah, and there is the wonderful type `STRINT_PAIR` - I suggest removing it on this occasion
(replacing all uses by its synonym `OPT_PAIR` - I tried it out, it works despite the comment in `opt.h`) - 
then there is no reason to document it :wink:

File: doc/internal/man3/OPTIONS.pod (hunk 21 above)
Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod:13]: For another PR for someone else to do.  What I wrote is good enough for a new openssl developer to write basic options.  (And `opt_int_arg` should crawl away to somewhere else :)

**Comment by richsalz** [comment]
Re: STRINT_PAIR:  speak to @paulidale about that one.

I maintain this is good enough to accept for internal documentation right now.  I'm not adding new content.

**Comment by richsalz** [comment]
fixup commit pushed to fix typo's.

File: doc/internal/man3/OPTIONS.pod (hunk 21 above)
Comment (DDvO) [reply doc/internal/man3/OPTIONS.pod:13]: Fine by me - the given coverage of functions should already be very helpful.

File: doc/internal/man3/OPTIONS.pod [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+        ...other options...
+        }
+    }
+
+The opt_help() function takes a list of option definitions and prints a
+nicely-formatted output.
+
+Within the option parsing loop, opt_flag() returns the option,
+without any leading hyphens. The opt_arg() function returns
+the option's value, if there is one..
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:222]: double period at EOL

File: doc/internal/man3/OPTIONS.pod [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+a directory, or B<0> if not. The opt_format() function takes a string value,
+such as used in with the B<-informat> or related flags, and fills
+the value from the constants in F<fmt.h> file.
+
+The opt_string() function checks that B<name> appears in the
+B<NULL>-terminated array of strings. It returns B<1> if found,
+or prints a diagnostic and returns B<0> if not.
+
+The opt_pair() function takes a list of B<pairs>, each of which
+has a text name and numeric value. The specified B<name> is
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:250]: numeric -> an integer

File: doc/internal/man3/OPTIONS.pod [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
+    R   Random state
+
+The B<OPT_x_ENUM> macro is used to define the numeration values, where B<x>
+is one of the letters above.  The B<OPT_x_OPTIONS> macro is used to
+list the set of common options, and the B<OPT_x_CASES> is used in
+the C<switch> statement.
+
+The common options are used throughout the sources for the OpenSSL commands.
+They are also used with common descriptions when generating the
+manpages, in the file C<doc/perlvars.pm>, which follow a similar naming
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:278]: C<doc/perlvars.pm> -> F<doc/perlvars.pm>

**Review by DDvO — CHANGES_REQUESTED** [review]
Just a few more nits, then I'll approve.

**Comment by richsalz** [comment]
updated commit pushed fixing all just-noticed typo's :)

**Review by DDvO — APPROVED** [review]
LGTM

File: doc/internal/man3/OPTIONS.pod [hunk 25]
```diff
@@ -0,0 +1,301 @@
...
+field is specified for the option. This means that error-checking inside
+the C<switch> C<case> can often be elided.
+
+The opt_isdir() function returns B<1> if the specified B<name> is
+a directory, or B<0> if not. The opt_format() function takes a string value,
+such as used in with the B<-informat> or related flags, and fills
+the value from the constants in F<fmt.h> file.
+
+The opt_string() function checks that B<name> appears in the
+B<NULL>-terminated array of strings. It returns B<1> if found,
```
Comment (t8m) [inline doc/internal/man3/OPTIONS.pod:246]: Remove `B<>` around NULL. Also I am not sure about bolding the constants such as 0 and 1.

File: doc/internal/man3/OPTIONS.pod [hunk 26]
```diff
@@ -0,0 +1,301 @@
...
+or prints a diagnostic and returns B<0> if not.
+
+The opt_pair() function takes a list of B<pairs>, each of which
```
Comment (t8m) [inline doc/internal/man3/OPTIONS.pod:249]: Parameters such as `pairs` here should be in `I<>` and not `B<>`.

**Review by t8m — CHANGES_REQUESTED** [review]
Please fix the markup.

File: doc/internal/man3/OPTIONS.pod (hunk 25 above)
Comment (richsalz) [reply doc/internal/man3/OPTIONS.pod:246]: Well it's not obvious :)
```
; g grep NULL doc | grep 'B<'|wc
    461    4893   43517
; g grep NULL doc | grep -v 'B<'|wc
   1362   11938  117044
```

But I removed the `B<>` around NULL and the 0/1 return values.

**Comment by richsalz** [comment]
fixup commit pushed to address the markup nits that @t8m found.

File: doc/internal/man3/OPTIONS.pod [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+    Valid options are:
+    ...detailed list of options...
+
+Sometimes there are parameters that should appear in the synopsis.
+Use B<OPT_HELP_STR> as the first entry in your array:
+
+    {OPT_HELP_STR, 1, '-', Usage: %s [options] [text...]\n"}
+
+The B<retval> and B<valtype> are ignored, and the B<helpstr> should
+follow the general construction as shown. The B<%s> will get the program
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:162]: C<%s>

File: doc/internal/man3/OPTIONS.pod [hunk 28]
```diff
@@ -0,0 +1,301 @@
...
+Within the option parsing loop, opt_flag() returns the option,
+without any leading hyphens. The opt_arg() function returns
+the option's value, if there is one.
+
+In an option list, there can be at most one option with the empty string.
+This is a "wildcard" or "unknown" option. For example, it allows an
+option to be be taken as digest algorithm, like B<-sha1>. The
+function opt_cipher() takes the specified B<name> and fills in
+the cipher into B<cipherp>.  The function opt_md() does the same
+thing for message digest.
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:229]: For all funciton arguments, throughout the doc, use I<> rather than B<>.

**Review by DDvO — CHANGES_REQUESTED** [review]
Some more markup nits.

**Comment by richsalz** [comment]
YAFCP  (yet another fixup commit pushed)

File: doc/internal/man3/OPTIONS.pod [hunk 29]
```diff
@@ -223,9 +223,9 @@ the option's value, if there is one.
 
 In an option list, there can be at most one option with the empty string.
 This is a "wildcard" or "unknown" option. For example, it allows an
-option to be be taken as digest algorithm, like B<-sha1>. The
-function opt_cipher() takes the specified B<name> and fills in
-the cipher into B<cipherp>.  The function opt_md() does the same
+option to be be taken as digest algorithm, like C<-sha1>. The
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:226]: YAN (yet another nit): 
B<-sha> was correct.

File: doc/internal/man3/OPTIONS.pod [hunk 30]
```diff
@@ -0,0 +1,301 @@
...
+The B<OPTIONS> C<typedef> specifies an option: what type of argument
+it takes (if any), and an optional "help" string.  It is a C<struct>
+containing these fields:
+
+    const char *name;
+    int retval;
+    int valtype;
+    const char *helpstr;
+
+The B<name> is the name of the option that the user would type. Options
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:91]: I<name>

File: doc/internal/man3/OPTIONS.pod [hunk 31]
```diff
@@ -0,0 +1,301 @@
...
+are words prefaced with a minus sign. If the user uses two minus signs,
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is the value to return if the option is found. It should be
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:96]: I<

File: doc/internal/man3/OPTIONS.pod [hunk 32]
```diff
@@ -0,0 +1,301 @@
...
+
+The opt_pair() function takes a list of I<pairs>, each of which
+has a text name and an integer. The specified I<name> is
+found on the list, it puts the index in I<*result>, and returns
+1. If not found, it returns 0.
+
+After processing all the options, the opt_num_rest() returns what is
+left, and opt_rest() returns a pointer to the first non-option.
+If there were no parameters, it will point to the NULL that is
+at the end of the standard B<argv> array.
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:257]: I<

File: doc/internal/man3/OPTIONS.pod [hunk 33]
```diff
@@ -0,0 +1,301 @@
...
+
+The B<name> is the name of the option that the user would type. Options
+are words prefaced with a minus sign. If the user uses two minus signs,
+this is also accepted for compatibility with other GNU software. Some
+names are special, and are described below.
+
+The B<retval> is the value to return if the option is found. It should be
+one of the choices in the enumeration above.
+
+The B<valtype> defines what the option's parameter must be. It should
```
Comment (DDvO) [inline doc/internal/man3/OPTIONS.pod:99]: I<

**Comment by richsalz** [comment]
I'm done.  Feel free to push your own fixup commits to this branch.

File: doc/internal/man3/OPTIONS.pod (hunk 30 above)
Comment (DDvO) [reply doc/internal/man3/OPTIONS.pod:91]: There are still many more such func arg names below.

**Review by DDvO — CHANGES_REQUESTED** [review]
The last fixup was incomplete.

**Comment by DDvO** [comment]
> YAFCP (yet another fixup commit pushed)

This would make a nice CLI command alias :rofl:

**Comment by richsalz** [comment]
> The last fixup was incomplete.

I encourage you to make the changes.  As I said, for this *INTERNAL* manpage I'm done.

**Comment by DDvO** [comment]
> > The last fixup was incomplete.
> 
> I encourage you to make the changes. As I said, for this _INTERNAL_ manpage I'm done.

Come on, just make the markup of function arg names consistent 
and I'll re-approve. And I presume @t8m will be happy as well.

**Comment by DDvO** [comment]
Maybe we should have a doc-nit-pick tool analogous to `check-format.pl` :wink:

**Comment by richsalz** [comment]
I made all the function arg names use `I<>`  You're now also asking for field names to be changed, if I understand comments like the one at line 96. And I shouldn't have to spend time trying to understand.  Shrug.  Change it, merge it, reject it. The nits have worn me down.

**Comment by DDvO** [comment]
> I made all the function arg names use `I<>` You're now also asking for field names to be changed, if I understand comments like the one at line 96. And I shouldn't have to spend time trying to understand. Shrug. 

Oops, right, this is an instance of a field name! For them I'm not sure what the convention is - @t8m?
I just know that, these days, func arg names are supposed to be written like this: `I<arg>`
since I was often asked to make changes in this direction.

`B<argv>` is a function arg name, so it should look like this: `I<argv>`.
Similarly, `"argc, argv"` would better be given as `I<argc>, I<argv>`.
And the CLI option `I<-sha1>` should be rendered as `B<-sha1>`.
Likely these four nits are the only ones remaining.

**Comment by t8m** [comment]
I'm willing to approve anyway as this is internal API doc.

**Review by t8m — APPROVED** [review]

**Review by DDvO — APPROVED** [review]
So I'm approving, too. 
Good enough for now; we can fix the remaining nits later.
Thanks @richsalz for taking up documenting `opt.h`.

**Comment by richsalz** [comment]
Thanks for the comments and feedback, glad this is getting merged.  Hope someone will fix all the remaining nits at some point, I guess.  (And yes, as the original author of find-doc-nits I appreciate the irony)

**Comment by DDvO** [comment]
Squashed and merged :)

**Comment by DDvO** [comment]
The issues identified above and not solved by this PR are now handled in the follow-up PR #15111.

**Pull request closed without merging**